File: packages/typescript/src/parser.ts
/**
 * Costco A-format price-tag parser. TypeScript port of the canonical Dart parser.
 * See ../../../spec/SPEC.md for the language-neutral algorithm this implements.
 */

import type { DiscountInfo, OcrLine, PriceTagData, TagType, UnitPrice } from "./types.js";

const PURE_ITEM = /^(\d{6})(\+?)$/;
// Digit boundaries so a 6-digit run inside a longer number (barcodes,
// "ITEM: 3663092") is NOT mistaken for an item number.
const ITEM = /(?<!\d)(\d{6})(?!\d)(\+?)/;
const PRICE = /(\d[\d,]{2,})\s*원/;
const COMMA_NUM = /\d{1,3}(?:,\d{3})+/;
// A comma-number followed by one of these is a nutrition/weight figure, not a price.
const WEIGHT_UNIT = /^\s*(?:kcal|kg|mg|ml|g|l)/i;
const DATE = /(\d{4})[.\/](\d{1,2})[.\/]?(\d{1,2})/g;
const UNIT = /단가\s*\/\s*([^\s]+)/;
const LATIN = /[A-Za-z]/;

interface PT {
  v: number;
  h: number;
  y: number;
}

const pad = (s: string | undefined): string => (s ?? "").padStart(2, "0");
const despace = (s: string): string => s.replace(/ /g, "");

/** ISO `YYYY-MM-DD`, or null when the date is out of calendar range. */
function isoDate(m: RegExpMatchArray): string | null {
  const y = +m[1];
  const mo = +m[2];
  const d = +m[3];
  if (y < 2000 || y > 2099 || mo < 1 || mo > 12 || d < 1 || d > 31) return null;
  return `${m[1]}-${pad(m[2])}-${pad(m[3])}`;
}

export function parsePriceTag(lines: OcrLine[]): PriceTagData {
  const raw = lines.map((l) => l.text).join("\n");
  const blob = lines.map((l) => l.text).join(" ");

  // --- itemNumber: prefer a line that is ONLY a 6-digit number (topmost) ---
  let item: string | null = null;
  let plus = false;
  // Stable reading order: top-to-bottom, then left-to-right. The `x` tiebreak
  // makes selection deterministic when two lines share yTop. See spec/SPEC.md.
  const byTop = (a: OcrLine, b: OcrLine): number => a.yTop - b.yTop || a.x - b.x;
  let pick: OcrLine | undefined = lines
    .filter((l) => PURE_ITEM.test(despace(l.text)))
    .sort(byTop)[0];
  if (!pick) {
    pick = lines.filter((l) => ITEM.test(despace(l.text))).sort(byTop)[0];
  }
  if (pick) {
    const m = ITEM.exec(despace(pick.text));
    if (m) {
      item = m[1];
      plus = m[2] === "+";
    }
  }

  // --- price tokens (value, glyph-height, vertical pos), in input order ---
  const pts: PT[] = [];
  for (const l of lines) {
    let v: number | null = null;
    const m = PRICE.exec(l.text);
    if (m) {
      v = parseInt(m[1].replace(/,/g, ""), 10);
    } else {
      const cm = COMMA_NUM.exec(l.text); // price whose '원' got split off
      // …unless it's a nutrition/weight figure (e.g. "2,142kcal"), not a price.
      if (cm && !WEIGHT_UNIT.test(l.text.slice(cm.index + cm[0].length))) {
        v = parseInt(cm[0].replace(/,/g, ""), 10);
      }
    }
    if (v !== null) pts.push({ v, h: l.h, y: l.yTop });
  }

  let finalTok: PT | null = null;
  for (const p of pts) {
    if (finalTok === null || p.h > finalTok.h) finalTok = p;
  }
  const finalPrice = finalTok ? finalTok.v : null;

  // --- tag type + discount ---
  const isDisc = blob.includes("할인행사");
  let tag: TagType;
  if (item === null && finalPrice === null) {
    tag = "unknown";
  } else {
    tag = isDisc ? "discount" : "regular";
  }

  let disc: DiscountInfo | null = null;
  if (tag === "discount") {
    let orig: number | null = null;
    if (finalPrice !== null) {
      for (const p of pts) {
        if (p.v > finalPrice && (orig === null || p.v > orig)) orig = p.v;
      }
    }
    const amt = orig !== null && finalPrice !== null ? orig - finalPrice : null;
    // Keep only calendar-plausible dates (OCR mangles them, e.g. 2026-06-70).
    const dates = [...blob.matchAll(DATE)].map(isoDate).filter((d): d is string => d !== null);
    const ps: string | null = dates.length > 0 ? dates[0] : null;
    const pe: string | null = dates.length > 1 ? dates[1] : null;
    disc = { originalPrice: orig, discountAmount: amt, periodStart: ps, periodEnd: pe };
  }

  // --- unit price: "단가 / <unit>" label + nearest small price ---
  let up: UnitPrice | null = null;
  let labelY: number | null = null;
  let unit: string | null = null;
  for (const l of lines) {
    const um = UNIT.exec(l.text);
    if (um) {
      labelY = l.yTop;
      unit = um[1];
      break;
    }
  }
  if (labelY !== null && unit !== null) {
    let best: PT | null = null;
    for (const p of pts) {
      if (finalTok !== null && p.v === finalTok.v && p.h === finalTok.h) continue; // skip final
      if (best === null || Math.abs(p.y - labelY) < Math.abs(best.y - labelY)) best = p;
    }
    if (best !== null) up = { value: best.v, unit };
  }

  // --- English name (best-effort): mostly-ASCII line with a space ---
  // Iterate CODE POINTS via spread, not .length (UTF-16 units).
  let nameEn: string | null = null;
  for (const l of lines) {
    const t = l.text.trim();
    const cps = [...t];
    if (cps.length < 5 || !t.includes(" ")) continue;
    const asciiPrintable = cps.filter((c) => {
      const o = c.codePointAt(0) ?? 0;
      return o > 32 && o < 128;
    }).length;
    if (asciiPrintable / cps.length > 0.7 && LATIN.test(t)) {
      nameEn = t;
      break;
    }
  }

  return {
    schemaVersion: "0.1",
    itemNumber: item,
    plusMark: plus,
    nameKo: null,
    model: null,
    nameEn,
    finalPrice,
    tagType: tag,
    discount: disc,
    unitPrice: up,
    rawText: raw,
  };
}

File: packages/python/src/costco_price_tag_parser/parser.py
"""Costco A-format price-tag parser. Pure Python port of the canonical Dart parser.

See ../../../spec/SPEC.md for the language-neutral algorithm this implements.
"""

from __future__ import annotations

import re

from .models import DiscountInfo, OcrLine, PriceTagData, TagType, UnitPrice

# Use explicit [0-9] (ASCII digits, like Dart/JS `\d`) instead of `\d` + re.ASCII —
# re.ASCII would ALSO make `\s` ASCII-only, diverging from Dart/JS where `\s` is
# Unicode (e.g. matches NBSP / full-width space between a price and 원).
_PURE_ITEM = re.compile(r"^([0-9]{6})(\+?)$")
# Digit boundaries so a 6-digit run inside a longer number (barcodes,
# "ITEM: 3663092") is NOT mistaken for an item number.
_ITEM = re.compile(r"(?<![0-9])([0-9]{6})(?![0-9])(\+?)")
_PRICE = re.compile(r"([0-9][0-9,]{2,})\s*원")
_COMMA_NUM = re.compile(r"[0-9]{1,3}(?:,[0-9]{3})+")
# A comma-number followed by one of these is a nutrition/weight figure, not a price.
_WEIGHT_UNIT = re.compile(r"^\s*(?:kcal|kg|mg|ml|g|l)", re.IGNORECASE)
_DATE = re.compile(r"([0-9]{4})[.\/]([0-9]{1,2})[.\/]?([0-9]{1,2})")
_UNIT = re.compile(r"단가\s*/\s*([^\s]+)")
_LATIN = re.compile(r"[A-Za-z]")


def _pad(s: str | None) -> str:
    return (s or "").rjust(2, "0")


def _iso_date(m: re.Match[str]) -> str | None:
    """ISO `YYYY-MM-DD`, or None when the date is out of calendar range (OCR
    routinely mangles dates, e.g. `2026.06.70`)."""
    y, mo, d = int(m.group(1)), int(m.group(2)), int(m.group(3))
    if not (2000 <= y <= 2099 and 1 <= mo <= 12 and 1 <= d <= 31):
        return None
    return f"{m.group(1)}-{_pad(m.group(2))}-{_pad(m.group(3))}"


def parse_price_tag(lines: list[OcrLine]) -> PriceTagData:
    raw = "\n".join(ln.text for ln in lines)
    blob = " ".join(ln.text for ln in lines)

    # --- itemNumber: prefer a line that is ONLY a 6-digit number (topmost) ---
    item: str | None = None
    plus = False
    pures = sorted(
        (ln for ln in lines if _PURE_ITEM.match(ln.text.replace(" ", ""))),
        key=lambda ln: (ln.y_top, ln.x),
    )
    pick = pures[0] if pures else None
    if pick is None:
        anys = sorted(
            (ln for ln in lines if _ITEM.search(ln.text.replace(" ", ""))),
            key=lambda ln: (ln.y_top, ln.x),
        )
        pick = anys[0] if anys else None
    if pick is not None:
        m = _ITEM.search(pick.text.replace(" ", ""))
        assert m is not None
        item = m.group(1)
        plus = m.group(2) == "+"

    # --- price tokens (value, glyph-height, vertical pos), in input order ---
    pts: list[tuple[int, float, float]] = []
    for ln in lines:
        v: int | None = None
        m = _PRICE.search(ln.text)
        if m:
            v = int(m.group(1).replace(",", ""))
        else:
            cm = _COMMA_NUM.search(ln.text)  # price whose '원' got split off
            # …unless it's a nutrition/weight figure (e.g. "2,142kcal"), not a price.
            if cm and not _WEIGHT_UNIT.match(ln.text[cm.end() :]):
                v = int(cm.group(0).replace(",", ""))
        if v is not None:
            pts.append((v, ln.h, ln.y_top))

    final_tok: tuple[int, float, float] | None = None
    for p in pts:
        if final_tok is None or p[1] > final_tok[1]:
            final_tok = p
    final_price = final_tok[0] if final_tok else None

    # --- tag type + discount ---
    is_disc = "할인행사" in blob
    if item is None and final_price is None:
        tag = TagType.UNKNOWN
    else:
        tag = TagType.DISCOUNT if is_disc else TagType.REGULAR

    disc: DiscountInfo | None = None
    if tag == TagType.DISCOUNT:
        orig: int | None = None
        if final_price is not None:
            for p in pts:
                if p[0] > final_price and (orig is None or p[0] > orig):
                    orig = p[0]
        amt = orig - final_price if (orig is not None and final_price is not None) else None
        # Keep only calendar-plausible dates (OCR mangles them, e.g. 2026-06-70).
        dates = [d for d in (_iso_date(m) for m in _DATE.finditer(blob)) if d]
        ps = dates[0] if dates else None
        pe = dates[1] if len(dates) > 1 else None
        disc = DiscountInfo(orig, amt, ps, pe)

    # --- unit price: "단가 / <unit>" label + nearest small price ---
    up: UnitPrice | None = None
    label_y: float | None = None
    unit: str | None = None
    for ln in lines:
        um = _UNIT.search(ln.text)
        if um:
            label_y = ln.y_top
            unit = um.group(1)
            break
    if label_y is not None and unit is not None:
        best: tuple[int, float, float] | None = None
        for p in pts:
            if final_tok is not None and p[0] == final_tok[0] and p[1] == final_tok[1]:
                continue  # skip final
            if best is None or abs(p[2] - label_y) < abs(best[2] - label_y):
                best = p
        if best is not None:
            up = UnitPrice(best[0], unit)

    # --- English name (best-effort): mostly-ASCII line with a space ---
    name_en: str | None = None
    for ln in lines:
        t = ln.text.strip()
        if len(t) < 5 or " " not in t:
            continue
        ascii_printable = sum(1 for c in t if 32 < ord(c) < 128)
        if ascii_printable / len(t) > 0.7 and _LATIN.search(t):
            name_en = t
            break

    return PriceTagData(
        item_number=item,
        plus_mark=plus,
        final_price=final_price,
        tag_type=tag,
        discount=disc,
        unit_price=up,
        name_en=name_en,
        raw_text=raw,
    )
